Zend/zend_compile.c

```diff
@@ -5158,6 +5158,244 @@ void zend_compile_switch(zend_ast *ast) /* {{{ */
+		if (arm_ast->child[0] == NULL) {
+			continue;
+		}
+
+		zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
+		num_conds += conds->children;
+	}
+
+	return num_conds;
+}
+
+static zend_uchar determine_match_jumptable_type(zend_ast_list *arms) {
+	zend_uchar common_type = IS_UNDEF;
+	for (uint32_t i = 0; i < arms->children; i++) {
+		zend_ast *arm_ast = arms->child[i];
+		if (!arm_ast->child[0]) {
+			/* Skip default clause */
+			continue;
+		}
+
+		zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
+		for (uint32_t j = 0; j < conds->children; j++) {
+			zend_ast **cond_ast = &conds->child[j];
+
+			zend_eval_const_expr(cond_ast);
+			if ((*cond_ast)->kind != ZEND_AST_ZVAL) {
+				/* Non-constant case */
+				return IS_UNDEF;
+			}
+
+			zval *cond_zv = zend_ast_get_zval(*cond_ast);
+			if (Z_TYPE_P(cond_zv) != IS_LONG && Z_TYPE_P(cond_zv) != IS_STRING) {
+				/* We only optimize switched on integers and strings */
+				return IS_UNDEF;
+			}
+
+			if (common_type == IS_UNDEF) {
+				common_type = Z_TYPE_P(cond_zv);
+			} else if (common_type != Z_TYPE_P(cond_zv)) {
+				/* Non-uniform case types */
+				return IS_UNDEF;
+			}
+		}
+	}
+
+	return common_type;
+}
+
+void zend_compile_match(znode *result, zend_ast *ast) /* {{{ */
+{
+	zend_bool result_unused = (ast->attr & ZEND_MATCH_RESULT_UNUSED) != 0;
+
+	zend_ast *expr_ast = ast->child[0];
+	zend_ast_list *arms = zend_ast_get_list(ast->child[1]);
+
+	if (result_unused || arms->children == 0) {
+		result->op_type = IS_CONST;
+		ZVAL_NULL(&result->u.constant);
+	}
+
+	zend_bool has_default_arm = 0;
+
+	uint32_t opnum_match = (uint32_t)-1;
+
+	znode expr_node;
+	zend_compile_expr(&expr_node, expr_ast);
+
+	zend_begin_loop(ZEND_FREE, &expr_node, 1);
+
+	znode case_node;
+	case_node.op_type = IS_TMP_VAR;
+	case_node.u.op.var = get_temporary_variable();
+
+	uint32_t num_conds = count_match_conds(arms);
+	zend_uchar jumptable_type = determine_match_jumptable_type(arms);
+	HashTable *jumptable = NULL;
+	if (jumptable_type != IS_UNDEF && should_use_jumptable(num_conds, jumptable_type)) {
+		znode jumptable_op;
+
+		ALLOC_HASHTABLE(jumptable);
+		zend_hash_init(jumptable, arms->children, NULL, NULL, 0);
+		jumptable_op.op_type = IS_CONST;
+		ZVAL_ARR(&jumptable_op.u.constant, jumptable);
+
+		zend_op *opline = zend_emit_op(NULL,
+			jumptable_type == IS_LONG ? ZEND_MATCH_LONG : ZEND_MATCH_STRING,
+			&expr_node, &jumptable_op);
+		if (opline->op1_type == IS_CONST) {
+			Z_TRY_ADDREF_P(CT_CONSTANT(opline->op1));
+		}
+		opnum_match = opline - CG(active_op_array)->opcodes;
+	}
+
+	uint32_t cond_count = 0;
+	uint32_t *jmpnz_opnums = safe_emalloc(sizeof(uint32_t), num_conds, 0);
+	for (uint32_t i = 0; i < arms->children; ++i) {
+		zend_ast *arm_ast = arms->child[i];
+
+		if (!arm_ast->child[0]) {
+			if (has_default_arm) {
+				CG(zend_lineno) = arm_ast->lineno;
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Match expressions may only contain one default arm");
```

Comment:
Nit: would be clearer as `may contain only one default arm`

Reply:
Yeah, this is copied from the switch so I'll leave it for now.